cmd/volume.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2017 XLAB, Ltd.
+ *
+ * This work is open source software, licensed under the terms of the
+ * BSD license as described in the LICENSE file in the top-level directory.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mikelangelo-project/capstan/hypervisor"
+	"github.com/mikelangelo-project/capstan/hypervisor/qemu"
+)
+
+// Volume is an extended version of hypervisor.Volume with volume creation data.
+type Volume struct {
+	hypervisor.Volume
+	SizeMB int64
+	Name   string
+}
+
+const VOLUMES_DIR string = "volumes"
+
+// CreateVolume creates volume with given specifications.
+func CreateVolume(packagePath string, volume Volume) error {
+	if _, err := os.Stat(filepath.Join(packagePath, "meta", "package.yaml")); os.IsNotExist(err) {
+		return fmt.Errorf("Must be in package root directory")
+	}
+	volumesDir := filepath.Join(packagePath, VOLUMES_DIR)
+	if _, err := os.Stat(volumesDir); os.IsNotExist(err) {
+		if err := os.Mkdir(volumesDir, 0777); err != nil {
+			return fmt.Errorf("Could not create volumes dir '%s': %s", volumesDir, err)
+		}
+	}
+
+	// Create actual volume.
+	volume.Path = filepath.Join(volumesDir, fmt.Sprintf("%s.%s", volume.Name, volume.Format))
+	if err := qemu.CreateVolume(volume.Path, volume.Format, volume.SizeMB); err != nil {
+		return fmt.Errorf("Could not create volume: %s", err)
+	}
+
+	// Write metadata.
+	if err := volume.PersistMetadata(); err != nil {
+		return fmt.Errorf("Could not persist volume metadata: %s", err)
+	}
+
+	return nil
+}
+
+// DeleteVolume deletes volume and its metadata with given name.
+func DeleteVolume(packagePath, name string) error {
+	volumesDir := filepath.Join(packagePath, VOLUMES_DIR)
+	if _, err := os.Stat(volumesDir); os.IsNotExist(err) {
+		return nil
+	}
+	name = fmt.Sprintf("%s.", name) // prevent deletion by same prefix -> "vol." is no longer prefix of "volume.raw"
+	err := filepath.Walk(volumesDir, func(path string, info os.FileInfo, err error) error {
```

Comment:
Why not check if the volume exists before attempting to remove corresponding file(s), e.g. with something similar as in L58? 🙂. I think it would make sense to return an error about a non-existing volume, just so that the user will get some feedback (perhaps she misspelled the name of the volume and now mistakenly thinks that it was deleted).